agent/api/task/task.go

```diff
@@ -1213,3 +1213,17 @@ func (task *Task) AddResource(resourceType string, resource taskresource.TaskRes
 	defer task.lock.Unlock()
 	task.ResourcesMapUnsafe[resourceType] = append(task.ResourcesMapUnsafe[resourceType], resource)
 }
+
+// Synchronize initializes the required field in the task on agent restart
+// Some of the fields in task isn't saved in the agent state file, agent needs
+// to initialize these fields before processing the task, eg: docker client in resource
+func (task *Task) Synchronize(resourceFields *taskresource.ResourceFields) {
+	task.lock.Lock()
+	defer task.lock.Unlock()
+
+	for _, resources := range task.ResourcesMapUnsafe {
+		for _, resource := range resources {
+			resource.Initialize(resourceFields)
```

Comment:
Is this going to be safe to reapply?

Should we make it clear that _all_ resources need to behave this way?

Reply:
yes, it's ok to reapply as it doesn't change any state. 
>Should we make it clear that all resources need to behave this way?

I think this depends on whether the resource needs to be initialized on restart.